Add mentor notes for csharp track nucleotide-count (#864)
exercism/website-copy: tracks/csharp/exercises/nucleotide-count/mentoring.md

### Reasonable solutions

> ### Using a foreach loop

Reviewer: Increase to level 4:
```suggestion
#### Using a foreach loop
```

Author: Resolved in 3bf9ce67cbd9ef4eef08e3718ec54ef4e11ed10c

```csharp
using System;
using System.Collections.Generic;
using System.Linq;

public static class NucleotideCount
{
    public static IDictionary<char, int> Count(string sequence)
    {

>         var result = new Dictionary<char, int> { {'A', 0}, {'C', 0}, {'G', 0}, {'T', 0} };

Reviewer 2: ```suggestion
        var nucleotideCounts = new Dictionary<char, int> { {'A', 0}, {'C', 0}, {'G', 0}, {'T', 0} };
```

Maybe use a more descriptive name?

Author: Resolved in 9fd8d58832c4d19e959cdf60c4d2429ec951bf12

        
        foreach (var c in sequence)
        {

>             if (result.ContainsKey(c)) result[c]++;

Reviewer 2: ```suggestion
            if (result.ContainsKey(c))
                result[c]++;
```

Having the statement of the if-clause on the same line is not something that one sees used often in C#.  Maybe put it on the next line (like [bob](https://github.com/exercism/website-copy/blob/master/tracks/csharp/exercises/bob/mentoring.md))?

Author: Happy to take it off the same line, but if so I'll use brackets because I find bracketless to be misleading and possibly dangerous (Some discussion of that here: https://softwareengineering.stackexchange.com/questions/145896/the-bracketless-if-statement-is-frowned-upon-but-is-it-still-bad-style-if-your)

Author: A good example of problems caused, in part, by bracketless IFs is Apple's 2014 SSL vulnerability: https://blog.codecentric.de/en/2014/02/curly-braces/

Reviewer 2: Sure, feel free to add brackets. That is most idiomatic.

Author: Resolved in 61a32fd4aefa4877c053be4d36cc1b7bffa696b8

>             else throw new ArgumentException("Invalid Nucleotide Symbol");

Reviewer 2: ```suggestion
            else
                throw new ArgumentException("Invalid Nucleotide Symbol");
```

Having the statement of the if-clause on the same line is not something that one sees used often in C#.  Maybe put it on the next line (like [bob](https://github.com/exercism/website-copy/blob/master/tracks/csharp/exercises/bob/mentoring.md))?

Author: Resolved in 61a32fd4aefa4877c053be4d36cc1b7bffa696b8

        }

        return result;
    }
}
```

> ### Foreach with ToDictionary

Reviewer: ```suggestion
#### Foreach with ToDictionary
```

Author: Resolved in 3bf9ce67cbd9ef4eef08e3718ec54ef4e11ed10c

```csharp
using System;
using System.Collections.Generic;
using System.Linq;

public static class NucleotideCount
{
    private static readonly char[] ValidNucleotideSymbols = {'A', 'C', 'G', 'T'};
    
    public static IDictionary<char, int> Count(string sequence)
    {

>         var result = ValidNucleotideSymbols.ToDictionary(c => c, c => 0);

Reviewer 2: ```suggestion
        var nucleotideCounts = ValidNucleotideSymbols.ToDictionary(c => c, c => 0);
```

Maybe use a more descriptive name?

Author: Good point, is more descriptive

Author: Resolved in 9fd8d58832c4d19e959cdf60c4d2429ec951bf12

        
        foreach (var c in sequence)
        {
            if (ValidNucleotideSymbols.Contains(c)) result[c]++;
            else throw new ArgumentException("Invalid Nucleotide Symbol");
        }

        return result;
    }
}
```

### Common Suggestions

>  * Iterating over `sequence` more than once wastes cycles, memory, etc. Often it isn't obvious that this is happening because methods like `ToList`, `GroupBy` or `OrderBy` will iterate a list without the student having to write a loop.

Reviewer 2: ```suggestion
 * Iterating over `sequence` more than once wastes cycles, memory, etc. Often it isn't obvious that this is happening because methods like `ToList`, `GroupBy` or `Count` will iterate a list without the student having to write a loop.
```

I think `Count` is something that will be used more often than `OrderBy`.

Author: Good point, will change

Author: Resolved in 44237bb40e87603671e8911323c2bf9b5fae5173

>  * Iterating over `sequence` more than once wastes cycles, memory, etc. Often it isn't obvious that this is happening because methods like `ToList`, `GroupBy` or `OrderBy` will iterate a list without the student having to write a loop.

Reviewer 2: ```suggestion
 * Iterating over the input string more than once wastes cycles, memory, etc. Often it isn't obvious that this is happening because methods like `ToList`, `GroupBy` or `OrderBy` will iterate a list without the student having to write a loop.
```

Author: Good point, they may change the incoming argument var name

Author: Resolved in 76ebad02da082fc3e34fd0bea3dee0437dee76b1

>  * Iterating over `sequence` more than once wastes cycles, memory, etc. Often it isn't obvious that this is happening because methods like `ToList`, `GroupBy` or `OrderBy` will iterate a list without the student having to write a loop.

Reviewer: Remove the list indicator here _or_ indent line [56](https://github.com/exercism/website-copy/pull/864/files#diff-6773c5801a8b5f93a7292c81cfadda3fR56) and [57](https://github.com/exercism/website-copy/pull/864/files#diff-6773c5801a8b5f93a7292c81cfadda3fR57).

The `Hence` and `Similarly` lines are suggestions based on this item.

Author: Lines 56 and 57 are further indented than 55 (two spaces), and display as sub-points in my markdown previewer.

Reviewer: You're right. I don't know what happened there -- displays fine on GH too.

   * Hence, using `sequence.Any()` or `sequence.Where(...)` to implement the exception is wrong. It should be thrown within the single iteration.

>    * Similarly, though there are numerous ways to use Linq to solve this problem it should be avoided. All of them involve iterating `sequence` multiple times.

Reviewer 2: ```suggestion
   * Similarly, though there are numerous ways to use LINQ to solve this problem it should be avoided. All of them involve iterating over the input string multiple times.
```

Author: Is "Linq" correct? MS use LINQ (https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/getting-started-with-linq), and refer to it as an acronym for Language INtegrated Query (https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/linq/).

Reviewer 2: Yes, that is correct, which is why I suggest to change it to "LINQ" :)

Author: Oh hah 😁 Apologies, I was reading the +/- backwards.

Reviewer 2: No problem :)

Author: Resolved in b917329ac775c006c52daf89fed1d9e411842b31
